src/net/server/handlers/mul.rs

```diff
@@ -0,0 +1,209 @@
+use crate::helpers::fabric::{ChannelId, MessageChunks, MessageEnvelope};
+use crate::helpers::Identity;
+use crate::net::server::MpcServerError;
+use crate::net::RecordHeaders;
+use crate::protocol::{QueryId, RecordId, UniqueStepId};
+use async_trait::async_trait;
+use axum::extract::{self, FromRequest, Query, RequestParts};
+use axum::http::Request;
+use axum::middleware::Next;
+use axum::response::Response;
+use hyper::Body;
+use tokio::sync::mpsc;
+
+/// Used in the axum handler to extract the `query_id` and `step` from the path of the request
+pub struct Path(QueryId, UniqueStepId);
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for Path {
+    type Rejection = MpcServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Path((query_id, step)) =
+            extract::Path::<(QueryId, UniqueStepId)>::from_request(req).await?;
+        Ok(Path(query_id, step))
+    }
+}
+
+#[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+pub struct IdentityQuery {
+    identity: Identity,
+}
+
+/// After an [`OwnedPermit`] has been reserved, it can be used once to send an item on the channel.
+/// Panics if cloned while a permit exists.
+pub struct ReservedPermit<T>(Option<mpsc::OwnedPermit<T>>);
+
+impl<T: Send + 'static> ReservedPermit<T> {
+    pub fn new(permit: mpsc::OwnedPermit<T>) -> Self {
+        Self(Some(permit))
+    }
+    /// # Panics
+    /// if called more than once
+    pub fn send(&mut self, item: T) {
+        self.0
+            .take()
+            .expect("should only call `send` once")
+            .send(item);
+    }
+}
+
+impl<T> Clone for ReservedPermit<T> {
+    /// # Panics
+    /// if a permit exists
+    fn clone(&self) -> Self {
+        assert!(self.0.is_none());
+        Self(None)
+    }
+}
+
+pub async fn message_stream_middleware<T: Send + 'static, B>(
```

Comment:
yea that is much better. I would also rename it to something like `obtain_permit_mw` because `message_stream` is not even a trait in the code.

Reply:
sounds good to me